### PR TITLE
LidoOracleNew, ValidatorExitBus: initial rough version of hash consensus

### PR DESCRIPTION
**Pull request by arwer13**
Initial untidy version of hash consensus, with further data report in single tx. Its purpose is to provide tooling with capability to test agains "two-phaseness".

It will be reimplemented/updated by further solid PR. 

**NB**: gas consumption of reporting keys to eject (call to `ValidatorExitBus.handleReportData`) is:
```
1: 121465 per key (121465 total)
3: 33941.666666666664 per key (101825 total)
10: 16263.9 per key (162639 total)
40: 10604.475 per key (424179 total)
100: 9519.35 per key (951935 total)
```
(see test "Calculate gas usages")

### CODE REVIEW

**Review by TheDZhon — COMMENTED** [review]
:rocket: Blazingly fast

Though, have some questions :eyes: :

File: contracts/0.8.9/ValidatorExitBus.sol [hunk 1]
```diff
@@ -139,33 +141,41 @@ contract ValidatorExitBus is CommitteeQuorum, AccessControlEnumerable, ReportEpo
...
+        bytes32 _reportHash
+    ) external {
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        bool hasEpochAdvanced = _validateAndUpdateExpectedEpoch(_epochId, beaconSpec);
+        if (hasEpochAdvanced) {
+            _clearReporting();
+        }
+
+        _handleMemberReport(msg.sender, _epochId, _reportHash);
+        // TODO: events
```
[the remark this answers is not in the document]
Comment (arwer13) [reply contracts/0.8.9/ValidatorExitBus.sol:147]: no events right here indeed

File: contracts/0.8.9/ValidatorExitBus.sol [hunk 2]
```diff
@@ -228,17 +238,17 @@ contract ValidatorExitBus is CommitteeQuorum, AccessControlEnumerable, ReportEpo
...
-            (
-                address[] memory stakingModules,
-                uint256[] memory nodeOperatorIds,
-                uint256[] memory validatorIds,
-                bytes[] memory validatorPubkeys,
-                uint256 epochId
-            ) = _decodeReport(distinctReports[reportIndex]);
-            _reportKeysToEject(stakingModules, nodeOperatorIds, validatorIds, validatorPubkeys, epochId, _getBeaconSpec());
-        }
+        // (bool isQuorumReached, uint256 reportIndex) = _updateQuorum(_quorum);
```
[the remark this answers is not in the document]
Comment (arwer13) [reply contracts/0.8.9/ValidatorExitBus.sol:233]: indeed!

**Comment by arwer13** [comment]
Current gas consumption of `ValidatorExitBus.handleReportData`:
```
1: 102572 per key (102572 total)
10: 13908.9 per key (139089 total)
100: 9133.84 per key (913384 total)
500: 9046.998 per key (4523499 total)
```

and without `emit ConsensusDataDelivered(…)`:
```
1: 97440 per key (97440 total)
10: 11227.8 per key (112278 total)
100: 6703.09 per key (670309 total)
500: 6457.42 per key (3228710 total)
```

File: contracts/0.8.9/CommitteeQuorum.sol [hunk 3]
```diff
@@ -80,66 +86,87 @@ contract CommitteeQuorum {
...
         uint256 i = 0;
         bool isFound = false;
-        while (i < distinctReports.length && distinctReportHashes[i] != reportHash) {
+        while (i < numDistinctReports && distinctReportHashes[i] != _reportHash) {
             ++i;
         }
-        while (i < distinctReports.length) {
-            if (distinctReportHashes[i] == reportHash) {
+        while (i < numDistinctReports) {
+            if (distinctReportHashes[i] == _reportHash) {
```
Comment (folkyatina) [inline contracts/0.8.9/CommitteeQuorum.sol:97]: IDK, check, please, but this `if` looks like always true for me. So while will run exactly once, if run at all.

**Review by folkyatina — APPROVED** [review]
Looks fine for our needs. 
There is some places that can be rewritten for clarity , but it is not worth it since we'll rewrite it anyway.

File: contracts/0.8.9/CommitteeQuorum.sol (hunk 3 above)
Comment (arwer13) [reply contracts/0.8.9/CommitteeQuorum.sol]: refactored, thanks

**Review by TheDZhon — APPROVED** [review]
It's good enough for testnets :+1:

**Pull request merged**